### ISSUE TITLE
Arity check for proc to include self break dry-type arrays

### ISSUE THREAD

**Issue by kbacha** [dry-rb/dry-initializer#33]
Upgrading to 2.3.0 while using arrays as input types to an initializer causes issues. This appears to be related to the way `self` is called when the arity is not 0. I think the intent was to allow a proc to contain a second argument and capture self.

Here's the discussion on the forum: https://discourse.dry-rb.org/t/interaction-between-di-2-3-and-dt/380
Here's an implementation that breaks: https://github.com/kbacha/types_and_initializer_bug

You can see the arity check here: https://github.com/dry-rb/dry-initializer/blob/7a605f2a5421747f99299b610fbbc85f801d75e3/lib/dry/initializer/builders/attribute.rb#L69-L73

Which ends up passing `self` to the `method` param in dry type arrays:
https://github.com/dry-rb/dry-types/blob/506b7588a7b776cd59e2e27538256472e4c8d3df/lib/dry/types/array/member.rb#L19-L21

**Comment by yewton** [comment]
FYI. Temporary workaround: `option :array, proc { |v| Types::Array.of(Types::Strict::String)[v] }`

**Comment by nepalez** [comment]
I remember the issue, but have no good solution yet

**Comment by bjeanes** [comment]
Oh boy, I just got bit by this but it took a while to track down. The level of `eval`-ing completely breaks `pry`'s ability to navigate what's going on so I really had to spend a lot of time poking and guessing what was going on here... 😞 

> I remember the issue, but have no good solution yet

@nepalez what about changing the method to something like:

``` ruby
def coercion_line
  return unless @type
  if @type.is_a?(Proc) && @type.arity.abs == 1
    "#{@val} = #{@item}.type.call(#{@val}, self) unless #{@null} == #{@val}"
  else
    "#{@val} = #{@item}.type.call(#{@val}) unless #{@null} == #{@val}"
  end
end
```

If not, is there are a use-case for passing self to a non-Proc?

**Comment by nepalez** [comment]
@bjeanes I'll look at the issue in a couple of days.

At the first glance, it seems acceptable, but usage of concrete type `Proc` is a hack. (tbh, usage of arity is a hack too).

**Comment by bjeanes** [comment]
I agree, on both counts. However, types quack sufficiently like procs in other ways (callable, etc) that there aren't obviously better solutions.

That being said, you could check for that `@type` is from `dry-types` explicitly, instead. That's at least a known value here and it would still allow other usages where another callable object could be passed in and benefit from the multi-arity behaviour...

Something like this, perhaps:

``` ruby
def coercion_line
  return unless @type
  if arity.abs == 1 || Dry::Types::Type === @type
     "#{@val} = #{@item}.type.call(#{@val}) unless #{@val} == #{@null}" 
   else 
     "#{@val} = #{@item}.type.call(#{@val}, self) unless #{@val} == #{@null}" 
   end 
end
```

I'm positive `Dry::Types::Type === @type` is not correct, but it is at least illustrative.

**Comment by nepalez** [comment]
@bjeanes Yes, I like the latest hack, because its more specific, than the previous one, but still quite general to accept subclasses of `Dry::Types::Type`

IMO even a better (more explicit) solution would be switching from the second argument to something like:

```ruby
"#{@val} = #{@item}.type.call(#{@val}, itself: self)
```

But this would require not only a new version of `dry-initializer`, but a corresponding update of `Dry::Types::Type` to accept second argument as an named, not a positional one. This seems too complicated 'cause `dry-types` is more popular than `dry-initializer` and its DSL should remain stable.

To recap: could you please provide a PR, I'd merged it asap, and updated a version.

**Comment by bjeanes** [comment]
@nepalez catching up on inbox here. I'm up for making a PR, but what I wasn't certain about in my original suggestion is whether `Dry::Types::Type === @type` is correct and works in all cases (I'm not familiar with dry-types internals.

**Comment by bjeanes** [comment]
It does seem to work for the array case and this issue is about in my local testing. I'll try to prepare a PR this week once I add a test for the change and then we can discuss who well it covers all dry-types scenarios on the associated PR#.
